eureka2-client/src/test/java/com/netflix/eureka2/client/interest/BatchAwareIndexRegistryTest.java

```diff
@@ -0,0 +1,165 @@
+
+    private final IndexRegistry<InstanceInfo> delegateIndexRegistry = mock(IndexRegistry.class);
+    private final BatchAwareIndexRegistry<InstanceInfo> indexRegistry =
+            new BatchAwareIndexRegistry<>(delegateIndexRegistry, remoteBatchingRegistry);
+
+    private final PublishSubject<ChangeNotification<InstanceInfo>> notificationSubject = PublishSubject.create();
+
+    private final ExtTestSubscriber<ChangeNotification<InstanceInfo>> testSubscriber = new ExtTestSubscriber<>();
+
+    @Before
+    public void setUp() throws Exception {
+        remoteBatchingRegistry.subscribe(remoteBatchingHintsSubject);
+
+        when(delegateIndexRegistry.forInterest(any(Interest.class), any(Observable.class), any(InitStateHolder.class)))
+                .thenReturn(notificationSubject);
+    }
+
+    /**
+     * Test situation where we subscribe for the first time for a particular
+     * interest, and there are two data items available in the server.
+     * <p>
+     * Expected behavior: two data notification followed by single finish buffering
+     */
+    @Test
+    public void testColdCacheAndDataInChannel() throws Exception {
+        indexRegistry.forInterest(ATOMIC_INTEREST_A, null, null).subscribe(testSubscriber);
+
+        sendRemoteBatch(ATOMIC_INTEREST_A, DATA_NOTIFICATION, DATA_NOTIFICATION);
+        verifyReceivedBatch(DATA_NOTIFICATION, DATA_NOTIFICATION);
+    }
+
+    /**
+     * Test situation where we subscribe second time to the same interest (2 items in cache), and there are
+     * no more data items available in the server.
+     * <p>
+     * Expected behavior: two data notification followed by single finish buffering
+     */
+    @Test
+    public void testHotCacheAndNoDataInChannel() throws Exception {
+        indexRegistry.forInterest(ATOMIC_INTEREST_A, null, null).subscribe(testSubscriber);
+        sendLocalBatch(ATOMIC_INTEREST_A, DATA_NOTIFICATION, DATA_NOTIFICATION);
+        verifyReceivedBatch(DATA_NOTIFICATION, DATA_NOTIFICATION);
+    }
+
+    /**
+     * Test situation where we subscribe second time to the same interest (2 items in cache), after the
+     * original subscriber disconnected (cold cache). There are two fresh data items on the server.
+     * <p>
+     * Expected behavior: two data notification followed by single finish buffering (from cache), followed by
+     * two data notification followed by single finish buffering (from server)
+     */
+    @Test
+    public void testStaleCacheAndDataInChannel() throws Exception {
+        indexRegistry.forInterest(ATOMIC_INTEREST_A, null, null).subscribe(testSubscriber);
+
+        sendLocalBatch(ATOMIC_INTEREST_A, DATA_NOTIFICATION, DATA_NOTIFICATION);
+        verifyReceivedBatch(DATA_NOTIFICATION, DATA_NOTIFICATION);
+
+        sendRemoteBatch(ATOMIC_INTEREST_A, DATA_NOTIFICATION, DATA_NOTIFICATION);
+        verifyReceivedBatch(DATA_NOTIFICATION, DATA_NOTIFICATION);
+    }
+
+    /**
```

Comment:
Description is different, but the expectation parts are identical, as we expect the same result.